File: backend/app/routers/huggingface.py
from fastapi import APIRouter

from app.models.huggingface import HfGgufFile, HfModelSummary
from app.services import huggingface_service

router = APIRouter(prefix="/api/huggingface", tags=["huggingface"])


@router.get("/search", response_model=list[HfModelSummary])
async def search(q: str, limit: int = 25) -> list[HfModelSummary]:
    return await huggingface_service.search_models(q, limit)


@router.get("/models/{repo_id:path}/files", response_model=list[HfGgufFile])
async def list_files(repo_id: str) -> list[HfGgufFile]:
    return await huggingface_service.list_gguf_files(repo_id)

File: app/src/api/huggingface.ts
import { api } from "./client";
import { HfGgufFile, HfModelSummary } from "./types";

export const huggingfaceApi = {
  search: (query: string) => api.get<HfModelSummary[]>(`/api/huggingface/search?q=${encodeURIComponent(query)}`),
  // repoId (e.g. "Qwen/Qwen2.5-Coder-7B-Instruct-GGUF") must keep its literal
  // "/" -- the backend route uses FastAPI's `{repo_id:path}` converter, which
  // expects real path segments, not a %2F-encoded single segment.
  listFiles: (repoId: string, serial?: string | null) =>
    api.get<HfGgufFile[]>(`/api/huggingface/models/${repoId}/files${serial ? `?serial=${serial}` : ""}`),
  download: (repoId: string, filename: string, deviceSerial?: string | null) =>
    api.post<{ job_id: string }>("/api/huggingface/download", { repo_id: repoId, filename, device_serial: deviceSerial }),
  cancel: (jobId: string) => api.post<{ status: string }>(`/api/huggingface/download/${jobId}/cancel`),
};
